### PR TITLE
Add option to deploy istio service mesh

### PR DESCRIPTION
**Pull request by rhrazdil**
Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

Added an option and a script to provision k8s cluster with Istio sevice mesh.

### CODE REVIEW

**Comment by rhrazdil** [comment]
To use istio with kubevirt, https://github.com/kubevirt/kubevirtci/pull/524 needs to be merged first
PR is ready for review, but putting on hold for now
/hold

cc @phoracek

File: cluster-provision/k8s/1.19/istio.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+set +e
+
+status=`kubectl --kubeconfig=/etc/kubernetes/admin.conf get pods --all-namespaces --no-headers --field-selector status.phase!=Running --output=name | grep istio`
+retry_counter=0
+while [[ $retry_counter -lt 30 && ! -z "$status" ]]; do
+    sleep 10
+    echo "Waiting for istio pods to be running..."
+    status=`kubectl --kubeconfig=/etc/kubernetes/admin.conf get pods --all-namespaces --no-headers --field-selector status.phase!=Running --output=name | grep istio`
+    retry_counter=$((retry_counter + 1))
+done
```
Comment (qinqon) [inline cluster-provision/k8s/1.19/istio.sh:21]: Doing this directly with golang inside gocli is a big no-no right ? It means we have to expose the apiserver port and all.

**Review by qinqon — CHANGES_REQUESTED** [review]
First pass, I don't see it enabled at ` check-cluster-up.sh`

File: cluster-provision/k8s/1.19/provision.sh [hunk 2]
```diff
@@ -2,12 +2,22 @@
 
 set -ex
 
-function docker_pull_retry() {
+KUBEVIRTCI_SHARED_DIR=/var/lib/kubevirtci
+mkdir -p $KUBEVIRTCI_SHARED_DIR
+cat << EOF > $KUBEVIRTCI_SHARED_DIR/common.sh
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply cluster-provision/k8s/1.19/provision.sh:7]: Done in https://github.com/kubevirt/kubevirtci/pull/541

File: cluster-provision/k8s/1.19/istio.sh (hunk 1 above)
Comment (rhrazdil) [reply cluster-provision/k8s/1.19/istio.sh:21]: Not sure tbh, I took inspiration from `ceph-csi.sh`. What apiserver port do you mean?

File: cluster-provision/k8s/1.19/provision.sh [hunk 3]
```diff
@@ -126,18 +115,14 @@ dnf install --skip-broken --nobest --nogpgcheck --disableexcludes=kubernetes -y
...
-cat <<EOT >/etc/sysconfig/kubelet
-KUBELET_EXTRA_ARGS=--cgroup-driver=systemd --runtime-cgroups=/systemd/system.slice --kubelet-cgroups=/systemd/system.slice --feature-gates="BlockVolume=true,CSIBlockVolume=true,VolumeSnapshotDataSource=true,IPv6DualStack=true"
-EOT
-
-systemctl daemon-reload
-
-systemctl enable docker && systemctl start docker
-systemctl enable kubelet && systemctl start kubelet
+cat <<EOF >/etc/sysconfig/kubelet
+KUBELET_EXTRA_ARGS=${KUBELET_CGROUP_ARGS} --feature-gates=${KUBELET_FEATURE_GATES}
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply cluster-provision/k8s/1.19/provision.sh:138]: You have a good point. I'm not sure what's the intention of these parameters added in nodes.sh: `--cpu-manager-policy=static --kube-reserved=cpu=500m --system-reserved=cpu=500m` maybe these should only be on worker nodes? 
But what certainly seems like a bug is that in nodes.sh the line is `appended`, so after nodes.sh runs, it contains 
```
KUBELET_EXTRA_ARGS="--cgroup-driver=systemd --runtime-cgroups=/systemd/system.slice --kubelet-cgroups=/systemd/system.slice --feature-gates=BlockVolume=true,CSIBlockVolume=true,VolumeSnapshotDataSource=true,IPv6DualStack=true,CPUManager=true"
KUBELET_EXTRA_ARGS=--cgroup-driver=systemd --runtime-cgroups=/systemd/system.slice --kubelet-cgroups=/systemd/system.slice --feature-gates=BlockVolume=true,CSIBlockVolume=true,VolumeSnapshotDataSource=true,IPv6DualStack=true,CPUManager=true --cpu-manager-policy=static --kube-reserved=cpu=500m --system-reserved=cpu=500m
```
I'll find out

[the remark this answers is not in the document]
Comment (rhrazdil) [reply cluster-provision/k8s/1.19/provision.sh:138]: I opended this PR to address the issue with KUBELET_EXTRA_ARGS put twice into the config file.
Also I have tried enablinch CPUManager and added this params that are not present in provision.sh: `--cpu-manager-policy=static --kube-reserved=cpu=500m --system-reserved=cpu=500m` and some pods fail to be scheduled, so I suppose we don't enable cpumanager on master node intentionally

**Review by qinqon — CHANGES_REQUESTED** [review]
Small nit

File: cluster-provision/k8s/check-cluster-up.sh [hunk 4]
```diff
@@ -24,6 +24,9 @@ export KUBEVIRTCI_GOCLI_CONTAINER=quay.io/kubevirtci/gocli:latest
   export KUBEVIRT_PROVIDER="k8s-${provision_dir}"
   export KUBEVIRT_NUM_NODES=2
   export KUBEVIRT_NUM_SECONDARY_NICS=2
+  if [[ $KUBEVIRT_PROVIDER =~ k8s-1\.2.* ]]; then
+    export KUBEVIRT_DEPLOY_ISTIO=true
+  fi
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply cluster-provision/k8s/check-cluster-up.sh]: makes sense, done

**Comment by rhrazdil** [comment]
Removed Istio deployment from check-cluster-up.sh because istio adds a configmap with a certificate to each new namespace.
For that reason, `should create a ResourceQuota and capture the life of a configMap` test fails. This test has [hardcoded requestquota for configMaps to 2](https://github.com/kubernetes/kubernetes/blob/02b9dd6a6e7d2928945afa4b8aedd7650d6d4cab/test/e2e/apimachinery/resource_quota.go#L1481), which doesn't take into account the additional one added by Istio.

@qinqon could you re-review please?

**Pull request merged**